File: Problems 1500-1599/1593. Split a String Into the Max Number of Unique Substrings/solution.py
class Solution:
    def maxUniqueSplit(self, s: str) -> int:
        def backtrack(s, start, seen):
            nonlocal count
            if start == len(s):
                count = max(count, len(seen))
                return

            # If we cannot split any more than current max, return
            if len(seen) + len(s) - start <= count:
                return

            for end in range(start, len(s)+1):
                substring = s[start:end + 1]
                if substring in seen:
                    continue
                seen.add(substring)
                backtrack(s, end + 1, seen)
                seen.remove(substring)
            return count

        count = 0
        backtrack(s, 0, set())
        return count

File: Problems 1500-1599/1593. Split a String Into the Max Number of Unique Substrings/solution.cpp
class Solution {
public:
    int maxUniqueSplit(string s) {
        unordered_set<string> seen;
        int maxCount = 0;
        return backtrack(s, 0, seen, maxCount);
    }

    int backtrack(string& s, int start, unordered_set<string>& seen, int& maxCount) {
        int n = s.length();
        if (start == n) {
            maxCount = max(maxCount, (int)seen.size());
            return maxCount;
        }

        if (seen.size() + n-start <= maxCount) return maxCount;

        for (int end = start; end < n; end++) {
            string split = s.substr(start, end-start+1);
            if (seen.find(split) != seen.end()) continue;
            seen.insert(split);
            backtrack(s, end+1, seen, maxCount);
            seen.erase(split);
        }
        return maxCount;
    }
};